lib/3scale/backend/transactor.rb

```diff
@@ -182,6 +188,24 @@ def preload_usage_limits(application, no_body_enabled, usage_in_params)
         end
       end
 
+      def filter_usages_with_0(transactions)
+        res = transactions.clone
```

Comment:
Why this call to `clone`?

Reply:
To avoid mutating the original `params[:usage]`, I think that would make the code and the tests harder to follow.

Reply:
But you're already mutating the original contents regardless of this - the clone is not a deep clone. The only reason I can think of to want to keep the original `params[:usage]` is for logging/monitoring, and at that point you can still let the callers decide to pass a clone or not. I don't think we are testing anywhere that the usage in params remains untouched anyway.

Reply:
Yes, I was thinking about aspects like logging. Would you be OK with this if I just change the call to `filter_usages_with_0(transactions.clone)` instead of cloning inside the function?

Reply:
Yes. But are we doing any such "aspects"?

Reply:
To answer that question you need to follow the code to be sure, and think about it every time you make a change in the callers chain. Right now that includes the few lines of code after the call to `filter_usages_with_0`, the caller method in `Listener` and all the rack middlewares that might run after it. I think it's easier to avoid mutating it.

Reply:
Ok. :+1:

Reply:
Changed this as explained above.